### ISSUE TITLE
write a new implementation of realtime sync for document editing

### ISSUE THREAD

**Issue by williamstein** [sagemathinc/smc#146]
There are some issues with realtime synchronized text editing on slower networks or when the project is on a heavily loaded machine so that connections drop. Mainly what happens is that it's not clear which patches got applied and which didn't, etc.     

I'll fix this by rewriting sync entirely to use a new approach based on a database-synchronized table using timestamps.   This is currently partly done in the file salvus/syncstring.coffee, but some work remains to switch it to being the default.

**Comment by atchekho** [comment]
Dear William, any updates on this? I am holding off on using Sagemath until this issue is resolved, as it was massively destructive last time it happened.  Thanks!

**Comment by williamstein** [comment]
No new updates, though I did write much of it and it is "in there", but off by default, since more work remains.  I had to do a massive refactor of everything to make SMC more development friendly, so somebody besides me can work on it.  I can't redo sync completely until that is done.  I just released most of that today, but a little more remains (e.g., to make it easy to fully work on SMC from within an SMC project).
